Initial shot at ClamAV 0.99.2

jimklimov opened the pull request:
Partially inspired by SFE manifest http://sfe.opencsw.org/localhostoi151a8/manifest/0/antivirus%2Fclamav@0.98,5.11-0.0.151.1.8%3A20131210T210011Z

jimklimov committed:
Initial commit of recipe for ClamAV 0.99.2

pyhalov commented on components/clamav/Makefile:
```diff
@@ -0,0 +1,76 @@
+### For X.Y.0 versions the TGZ and DIR do not include trailing .0 piece
+COMPONENT_VERSION_DIR=	0.99
+COMPONENT_VERSION=	$(COMPONENT_VERSION_DIR).0
+#COMPONENT_VERSION=	$(COMPONENT_VERSION_DIR)
```
Please, remove unnecessary commented out code

jimklimov replied:
That 'unnecessary code' will become needed for non-.0 versions and vice
versa.

## 

Typos courtesy of GMail on my Samsung Android
05.02.2016 12:22 пользователь "Alexander Pyhalov" notifications@github.com
написал:

> In components/clamav/Makefile
> https://github.com/OpenIndiana/oi-userland/pull/1684#discussion_r52004516
> :
> 
> > +# A full copy of the text of the CDDL should have accompanied this
> > +# source. A copy of the CDDL is also available via the Internet at
> > +# http://www.illumos.org/license/CDDL.
> > +#
> > +
> > +#
> > +# Copyright 2016, Jim Klimov. All rights reserved.
> > +#
> > +
> > +include ../../make-rules/shared-macros.mk
> > +
> > +COMPONENT_NAME=        clamav
> > +### For X.Y.0 versions the TGZ and DIR do not include trailing .0 piece
> > +COMPONENT_VERSION_DIR= 0.99
> > +COMPONENT_VERSION= $(COMPONENT_VERSION_DIR).0
> > +#COMPONENT_VERSION=    $(COMPONENT_VERSION_DIR)
> 
> Please, remove unnecessary commented out code
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OpenIndiana/oi-userland/pull/1684/files#r52004516.

jimklimov replied:
fixed

pyhalov commented on components/clamav/Makefile:
```diff
@@ -0,0 +1,76 @@
+include $(WS_TOP)/make-rules/configure.mk
+include $(WS_TOP)/make-rules/ips.mk
+
+PKG_OPTIONS += \
```
please, split this and CONFIGURE_OPTIONS lists line-by line

jimklimov replied:
fixed

pyhalov commented on components/clamav/Makefile:
```diff
@@ -0,0 +1,76 @@
+	-D COMPONENT_VERSION_LIB_FULL="$(COMPONENT_VERSION_LIB_FULL)" \
+	-D COMPONENT_VERSION_LIB_TDL="$(COMPONENT_VERSION_LIB_TDL)"
+
+CONFIGURE_OPTIONS +=	--prefix=/usr --sysconfdir=/etc
```
--prefix is set by default

jimklimov replied:
fixed

pyhalov commented on components/clamav/clamav-clamdtop.p5m:
```diff
@@ -0,0 +1,60 @@
+# CDDL HEADER END
+#
+# Copyright (c) 2016 Jim Klimov. All rights reserved.
+# Partially inspired by SFE manifest
```
I think this text should go to commit message, not to the file

jimklimov replied:
What do you mean? the "partially inspired" line? other "text" is standard, more or less..

pyhalov replied:
yes

jimklimov replied:
fixed

pyhalov replied:
Still see the text

pyhalov commented on components/clamav/clamav-clamdtop.p5m:
```diff
@@ -0,0 +1,60 @@
+### The above line should ensure common ClamAV dependencies du-jour
+### as well as "clamav" UID and GID
+
+#depend fmri=library/pcre type=require
```
why these lines are commented out?

jimklimov replied:
Initially, dependencies were copied from that SFE manifest, but then pkgmogrify said that these are conflicting errors as it detected equivalent deps. In the end, they all got commented away just to get this package built and published into my repo.

jimklimov replied:
fixed

pyhalov commented on components/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,196 @@
+dir group=clamav owner=clamav mode=0755 path=var/clamav/run
+
+file \
+    files/var/svc/manifest/site/clamav.xml \
```
we usually don't use such hierarchy for files/ . Please, put files to files/ .

jimklimov replied:
fixed

pyhalov commented on components/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,196 @@
+    path=usr/bin/freshclam.sh \
+    group=bin owner=root mode=0555
+
+file \
```
Perhaps, it's better to put examples to /usr/share/clamav or /usr/share/doc/clamav ?

jimklimov replied:
Possibly better, though the clamav-enable.sh script does simplify initial setup by copying adapted examples (*.sol files) as active configs. Can copy from anywhere though.

jimklimov replied:
fixed

pyhalov commented on components/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,196 @@
+    path=etc/freshclam.conf.sol \
+    group=sys mode=0444 owner=root
+
+file \
```
what is the origin of this file?

jimklimov replied:
which file (maybe i can't see some markup on phone-mailer)?

the .sol configs are my older defaults for the package I maintained
in-house (pre-set pathnames and user accounts, mostly). the .sh scripts are
mine.

## 

Typos courtesy of GMail on my Samsung Android
05.02.2016 12:33 пользователь "Alexander Pyhalov" notifications@github.com
написал:

> In components/clamav/clamav-core.p5m
> https://github.com/OpenIndiana/oi-userland/pull/1684#discussion_r52005619
> :
> 
> > -    path=etc/clamd.conf.sample-v$(COMPONENT_VERSION) \
> > -    group=sys mode=0444 owner=root
> >   +file \
> > -    files/etc/clamd.conf.sol \
> > -    path=etc/clamd.conf.sol \
> > -    group=sys mode=0444 owner=root
> >   +file \
> > -    etc/freshclam.conf.sample \
> > -    path=etc/freshclam.conf.sample-v$(COMPONENT_VERSION) \
> > -    group=sys mode=0444 owner=root
> >   +file \
> > -    files/etc/freshclam.conf.sol \
> > -    path=etc/freshclam.conf.sol \
> > -    group=sys mode=0444 owner=root
> >   +
> >   +file \
> 
> what is the origin of this file?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OpenIndiana/oi-userland/pull/1684/files#r52005619.

pyhalov commented on components/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,196 @@
+    path=usr/share/clamav/clamav-test.tgz \
+    mode=0444 owner=root group=sys
+
+file \
```
I don't think we have to ship *.log files as actions. If they are really needed, service could create them on startup.

jimklimov replied:
pkgmogrify (or someone in pkg stack) refuses to make symlinks to absent objects

jimklimov replied:
fixed (removed the empty logs and symlinks)

pyhalov commented on components/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,196 @@
+dir path=usr/share/clamav mode=0755 owner=root group=sys
+file path=usr/include/clamav.h mode=0444 owner=root group=sys
+
+file path=usr/share/man/man5/freshclam.conf.5 mode=0444 owner=root group=sys
```
permissions for most files should be set automatically by transforms

jimklimov replied:
fixed

pyhalov commented on components/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,196 @@
+hardlink path=usr/bin/freshclam target=../lib/isaexec pkg.linted=true
+hardlink path=usr/sbin/clamd target=../lib/isaexec pkg.linted=true
+
+file path=usr/lib/libclamunrar_iface.la mode=0555 owner=root group=sys
```
we don't deliver *.la files

jimklimov replied:
fixed

pyhalov commented on components/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,196 @@
+link path=usr/lib/$(MACH64)/libclamav.so target=libclamav.so.$(COMPONENT_VERSION_LIB_FULL)
+link path=usr/lib/$(MACH64)/libclamunrar.so.$(COMPONENT_VERSION_LIB_MAJOR) target=libclamunrar.so.$(COMPONENT_VERSION_LIB_FULL)
+
+### If --enable-libltdl-install, conflicts with libtool
```
remove commented out code

jimklimov replied:
ok. again, initially built with this like in-house, then discovered that the OS has the library ;)

jimklimov replied:
fixed

pyhalov commented on components/clamav/clamav-milter.p5m:
```diff
@@ -0,0 +1,83 @@
+    path=/var/clamav/log/clamav-milter.log \
+    mode=0644 owner=clamav group=clamav preserve=custom
+
+link path=var/log/clamav-milter.log target=../clamav/log/clamav-milter.log
```
don't do this, think about people, trying to rotate log file

jimklimov replied:
fixed (removed the empty logs and symlinks; made a /var/log/clamav/ symlink to directory instead)

pyhalov commented on components/clamav/files/etc/clamav-milter.conf.example-v0.98.1:
```diff
@@ -0,0 +1,293 @@
+##
```
Why can't we use example from package, why should we store it? If necessary, it can be patched. Delivering own file can be suitable if patch is too big.

jimklimov replied:
Sorry, old _standard_ config sample files slipped in from in-house legacy ;)
Still, they did not change in upstream sources since the specified versions.

Fixed away.

pyhalov commented on components/clamav/files/usr/bin/clamav-enable.sh:
```diff
@@ -0,0 +1,54 @@
+#!/bin/sh
```
Do we really need it? Better to deliver example files as /etc/*.conf

jimklimov replied:
Not always that we install the package we actually want the service activated (e.g. usability starts with tens of megabytes of virusdb download - not something I'd want forced onto me when I'm on cellular internet, and maybe from every zone that has a copy installed for whatever reason such as testing-staging, but could share the same db with some sysadmin creativity). Also perhaps I don't want the daemons running at all and a CLI tool suffices for a certain installation.

pyhalov commented on components/clamav/files/var/svc/manifest/site/freshclam.xml:
```diff
@@ -0,0 +1,36 @@
+<?xml version='1.0'?>
```
perhaps, do something like time-slider SMF-manged cron tasks?

jimklimov replied:
If I can ensure that a next iteration will be scheduled (despite any intermittent errors), and two iterations won't race in conflict at the same time - why not. As a first shot that gets the job done, a loop which sequentially updates the database and can be stopped/started as an smf service seemed sufficient until proven otherwise. (Might need corrections about service type though, e.g. transient maybe...)

jimklimov replied:
fixed (I hope)

jimklimov commented:
Fixed many of the comments (all that I planned to fix), now running a local re-publish to make sure this is still intact enough for gmake and pkg(5) toolkits ;)

jimklimov commented:
Ok, this state of things built and installed successfully. I am a bit worried about these lines in the end of the log though:

```
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/bin/amd64/clamconf' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/bin/amd64/clamscan' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/bin/amd64/clamsubmit' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/bin/amd64/sigtool' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/bin/amd64/clambc' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/bin/amd64/freshclam' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/bin/amd64/clamdscan' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/sbin/amd64/clamd' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/lib/amd64/libclamunrar_iface.so.7.1.1' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/lib/amd64/libclamunrar.so.7.1.1' includes '/usr/lib'
WARNING userland.action001.3      32-bit runpath in 64-bit binary, '/home/jim/shared/oi-userland/components/clamav/build/prototype/i386/usr/lib/amd64/libclamav.so.7.1.1' includes '/usr/lib'
```

Otherwise, the amd64 bins called explicitly do work as expected, and there are no resolved 32-bit libs, e.g.

```
root@nbofh:~# ldd /usr/bin/amd64/clamscan 
    libclamav.so.7 =>    /usr/lib/amd64/libclamav.so.7
    libsocket.so.1 =>    /usr/lib/amd64/libsocket.so.1
    libpthread.so.1 =>   /usr/lib/amd64/libpthread.so.1
    libc.so.1 =>     /usr/lib/amd64/libc.so.1
    libxml2.so.2 =>  /usr/lib/64/libxml2.so.2
    libbz2.so.1 =>   /usr/lib/64/libbz2.so.1
    libltdl.so.7 =>  /usr/lib/64/libltdl.so.7
    libssl.so.1.0.0 =>   /lib/64/libssl.so.1.0.0
    libcrypto.so.1.0.0 =>    /lib/64/libcrypto.so.1.0.0
    libz.so.1 =>     /usr/lib/64/libz.so.1
    libpcre.so.0 =>  /usr/lib/64/libpcre.so.0
    libm.so.2 =>     /lib/64/libm.so.2
    libnsl.so.1 =>   /lib/64/libnsl.so.1
    liblzma.so.5 =>  /usr/lib/64/liblzma.so.5
    libdl.so.1 =>    /lib/64/libdl.so.1
    libmp.so.2 =>    /lib/64/libmp.so.2
    libmd.so.1 =>    /lib/64/libmd.so.1
root@nbofh:~# ldd /usr/bin/amd64/clamscan  | grep -v 64
```

Any idea how to fix (and if this should be addressed at all)?

jimklimov commented:
For the record, I've also contemplated if the default SMF manifest for `clamav-milter` should include any sort of dependency with `sendmail` instances (e.g. if both are enabled - (re)start them in specific order), but chose against such complexity at least for initial integration.
Also I have examples of non-sendmail servers with milter interface that talk to ClamAV, and even that's on different hosts, so local SMF dependencies were not helpful, neither with sendmail nor an actual custom email service.

pyhalov commented:
As for "32-bit runpath in 64-bit binary" - something adds -R /usr/lib to LDFLAGS. It is not critical, but it can be interesting to find out, where  it is coming from.

jimklimov commented:
According to logs and resulting Makefiles - indeed, LDFLAGS is populated with a `${libdir}` component which is defined (on command line and variable dumps) as `/usr/lib/amd64` - but the result involves `/usr/lib` anyway.
Some bug in ClamAV `configure` script I'd guess.

Beside this, is it now good to rebase (and take into account that comment) to go into the PR as one commit?

jimklimov commented:
Tried to ensure creation of `configure` script with a newer (oi-hipster) version of autotools with

```
+COMPONENT_PREP_ACTION = \
+       (cd $(@D); \
+               libtoolize --copy --force ; \
+               aclocal -I . ; \
+               autoheader ; \
+               automake --add-missing --copy ; \
+               autoconf)
+
```

but result is the same.

pyhalov commented on components/clamav/files/freshclam.xml:
```diff
@@ -0,0 +1,36 @@
+    <instance name='default' enabled='false'>
+      <exec_method type='method'
+                   name='start'
+                   exec='while true ; do /usr/bin/freshclam.sh &amp;&amp; echo OK || echo FAILED ; sleep 900; echo Refreshing ClamAV database; done'
```
This still seems to be a hack. Can start/stop method just operate on clamav crontab?

jimklimov replied:
Hopefully made better by a recent commit. Review and testing is advised ;)
Now that this is not a loop, nothing forbids this from queuing up in case of some long lags or slow downloads; however the cron schedule (enabled upon every service startup, removed upon every shutdown) should be configurable as a variable in the SMF service so particular installations might have a way to amend this ;)
Also note that the packages currently do not depend on cron being installed (user can quite run freshclam manually when wanted). The SMF manifest would likely fail to enable, and the method script would check for "crontab" in PATH as well as during execution and error out if this fails. On the upside, this should be relatively agnostic of the cron version and implementation (not tested though).

jimklimov commented:
Fixed and rebuilt and reinstalled and fixed and...
Now works for me without hiccups, so far :-)

pyhalov commented:
downloading... I/O Error: Can't open url http://www.clamav.net/downloads/production/clamav-0.99.tar.gz: 
HTTP Error 403: Forbidden

jimklimov commented:
Strange... works for me via CLI wget, even through proxy : 

clamav$ wget http://www.clamav.net/dow
--2016-02-24 11:59:02--  http://www.clamav.net/downloads/production/clamav-0.99.tar.gz
Resolving proxy.localdomain (proxy.localdomain)... x.y.z.a
Connecting to proxy.localdomain (proxy.localdomain)|x.y.z.a|:8080... connected.
Proxy request sent, awaiting response... 302 Found
Location: https://s3.amazonaws.com/clamav-site/production/release_files/files/000/000/250/original/clamav-0.99.tar.gz?AWSAccessKeyId=AKIAIY6OSGQFGUNJQ7GQ&Expires=1456315148&Signature=oBd2fI%2B%2BexmRWrv1md8ugWLz7sc%3D [following]
--2016-02-24 11:59:08--  https://s3.amazonaws.com/clamav-site/production/release_files/files/000/000/250/original/clamav-0.99.tar.gz?AWSAccessKeyId=AKIAIY6OSGQFGUNJQ7GQ&Expires=1456315148&Signature=oBd2fI%2B%2BexmRWrv1md8ugWLz7sc%3D
Connecting to proxy.localdomain (proxy.localdomain)|x.y.z.a|:8080... connected.
Proxy request sent, awaiting response... 200 OK
Length: 15968038 (15M) [binary/octet-stream]
Saving to: ‘clamav-0.99.tar.gz’

clamav-0.99.tar.gz   35%[======>               ]   5.38M   493KB/s   eta 40s

However "gmake download" does fail too, instantly in the "userland-fetch" step. Maybe some issue with AWS, or HTTPS, or... ?

mtelka commented:
I had similar problem with downloading hexchat.  Maybe the root cause is same here.  I had to change the User-Agent during the download to workaround the problem.

For details look at these commits: d9716386dc0eb30f81b90eac6e4cf889f1cd9a02 and a9be2dae309b16685f70db3dc19faa9d0198caca

jimklimov commented:
Thanks a lot Marcel, your neat trick did the trick ;)
At least, "gmake download" now succeeds for me.

pyhalov commented on components/clamav/Makefile:
```diff
@@ -0,0 +1,75 @@
+COMPONENT_ARCHIVE_URL=	\
+	http://www.clamav.net/downloads/production/$(COMPONENT_ARCHIVE)
+# We need to use non-default User-Agent for the fetch, otherwise we are denied
+# to download the source file. (Thanks to Marcel Telka)
```
Please, remove gratitude from comment. You'd better thank him personally :)

jimklimov replied:
done

pyhalov commented on components/clamav/Makefile:
```diff
@@ -0,0 +1,75 @@
+CONFIGURE_OPTIONS +=	--with-xml=/usr
+
+### The clamav-milter requires that libmilter exists (in OI Hipster's
+### sendmail-8.14.4, that is only present for 32-bits it seems...)
```
Please, rephrase: Milter is enabled only in 32-bit build as illumos-gate currently doesn't provide 64-bit libmilter.
Also it would be good to file feature request against illumos-gate.

pyhalov commented on components/clamav/clamav-clamdtop.p5m:
```diff
@@ -0,0 +1,53 @@
+
+file path=usr/share/man/man1/clamdtop.1
+
+### Reserved for the case we'd have a 64-bit ncurses (pdcurses) later
```
We do have 64-bit ncurses, what exactly is missing?

jimklimov replied:
Seems that the Makefile should've defined `LDFLAGS.$(BITS)` to `-L` and `-R` proper libdir variants. Verifying locally...

pyhalov commented on components/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,149 @@
+# CDDL HEADER END
+#
+# Copyright (c) 2016 Jim Klimov. All rights reserved.
+# Partially inspired by SFE manifest
```
please, remove this text

pyhalov commented on components/clamav/files/freshclam.xml:
```diff
@@ -0,0 +1,45 @@
+    <instance name='default' enabled='false'>
+      <exec_method type='method'
+                   name='start'
+                   exec='/usr/bin/freshclam.sh ; /usr/bin/freshclam.sh --add-cron "%{freshclam/cron-schedule}"'
```
it seems, something is wrong:
/usr/bin/freshclam.sh --add-cron "0,15,30,45\ *\ *\ *\ *" doesn't work here:
$ pfexec svcadm enable freshclam
$ pfexec crontab -l clamav
crontab: can't open your crontab file.

jimklimov replied:
If you had the user created by previous dev-version, it was NP not NL - and
those are blocked from crontab. See recent commits (change to freshclam
smf) for details ;)

## 

Typos courtesy of GMail on my Samsung Android
25.02.2016 6:41 пользователь "Alexander Pyhalov" notifications@github.com
написал:

> In components/clamav/files/freshclam.xml
> https://github.com/OpenIndiana/oi-userland/pull/1684#discussion_r54051074
> :
> 
> > +<service_bundle type='manifest' name='clamav'>
> > -  <service name='antivirus/freshclam' type='service' version='0'>
> > -    <single_instance/>
> > -    <dependency name='fs' grouping='require_all' restart_on='none' type='service'>
> > -      <service_fmri value='svc:/system/filesystem/local'/>
> > -    </dependency>
> > -    <dependency name='network' grouping='require_all' restart_on='none' type='service'>
> > -      <service_fmri value='svc:/milestone/multi-user-server'/>
> > -    </dependency>
> > -    <dependency name='cron' grouping='require_all' restart_on='none' type='service'>
> > -      <service_fmri value='svc:/system/cron:default'/>
> > -    </dependency>
> > -    <instance name='default' enabled='false'>
> > -      <exec_method type='method'
> > -                   name='start'
> > -                   exec='/usr/bin/freshclam.sh ; /usr/bin/freshclam.sh --add-cron "%{freshclam/cron-schedule}"'
> 
> it seems, something is wrong:
> /usr/bin/freshclam.sh --add-cron "0,15,30,45\ *\ *\ *\ *" doesn't work
> here:
> $ pfexec svcadm enable freshclam
> $ pfexec crontab -l clamav
> crontab: can't open your crontab file.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OpenIndiana/oi-userland/pull/1684/files#r54051074.

pyhalov replied:
no, it's not the case. Something seems to be wrong with smf startup method

jimklimov commented:
While the train was enroute, the puppy grew up - bumping to 0.99.1 soon...

jimklimov commented:
Collapsed into one commit.

jimklimov commented:
Re-tested by removing an old installation, installing fresh result of `gmake publish` and running the enabling script to download the virus definitions.

Removing:

```
root@jimoi:~# pkg uninstall -v clamav{,-{core,milter,clamdtop}}
            Packages to remove:         4
            Services to change:         1
     Estimated space available:   7.29 GB
Estimated space to be consumed: 626.17 MB
       Create boot environment:        No
Create backup boot environment:        No
          Rebuild boot archive:        No

Changed packages:
oi-userland-jim
  antivirus/clamav
    0.99.1-2015.0.2.0 -> None
  antivirus/clamav-clamdtop
    0.99.1-2015.0.2.0 -> None
  antivirus/clamav-core
    0.99.1-2015.0.2.0 -> None
  antivirus/clamav-milter
    0.99.1-2015.0.2.0 -> None

Services:
  restart_fmri:
    svc:/system/manifest-import:default

PHASE                                          ITEMS
Removing old actions                         169/169
Updating package state database                 Done 
Updating package cache                           4/4 
Updating image state                            Done 
Creating fast lookup database                   Done 
Updating package cache                           3/3 

root@jimoi:~# find /var/svc/manifest/ | grep clam
root@jimoi:~# svcs -a | grep clam
root@jimoi:~# ls -la /var/spool/cron/crontabs/*clam*
ls: cannot access /var/spool/cron/crontabs/*clam*: No such file or directory

### Custom configs remain, arguably not a bug - they are (may be) customized by local admin
root@jimoi:~# ls -la /etc/*clam*
-rw-r--r-- 1 root root  8831 Feb 22 21:27 /etc/clamav-milter.conf
-rw-r--r-- 1 root root 15485 Feb 22 21:27 /etc/clamd.conf
-rw-r--r-- 1 root root  8090 Feb 22 21:27 /etc/freshclam.conf

### So I just move mine away for clarity of experiment - things did succeed
### in a previous experiment when I did not move these files away...
root@jimoi:~# mv /etc/*clam* /var/tmp

### Also a curious case logged in freshclam SMF service:
[ Mar  9 13:57:40 Stopping because service disabled. ]
[ Mar  9 13:57:40 Could not interpret "user" property value "clamav", error 2. ]
```

I guess there is also what we can name a bug in un-installation approach:

```
### Uninstallation is not completely clean - the group remains...
root@jimoi:~# grep clam /etc/{passwd,group,shadow}
/etc/group:clamav::11:

### ...ok - maybe this is a semi-valid reason - the downloaded databases and the logs are not killed off
### (no idea how to do this without pre-uninstall ;) and if it should be done in case the user "just upgrades" etc. -
### however the /var/clamav/db is thus removed by pkg(5) so we do not reuse these downloads anyway):
root@jimoi:~# find / -group 11 2>/dev/null
/var/pkg/lost+found/var/clamav/db-20160309T135657Z
/var/pkg/lost+found/var/clamav/db-20160309T135657Z/safebrowsing.cvd
/var/pkg/lost+found/var/clamav/db-20160309T135657Z/mirrors.dat
/var/pkg/lost+found/var/clamav/db-20160309T135657Z/daily.cvd
/var/pkg/lost+found/var/clamav/db-20160309T135657Z/main.cvd
/var/pkg/lost+found/var/clamav/db-20160309T135657Z/bytecode.cvd
/var/pkg/lost+found/var/clamav/db-20160308T161439Z
/var/pkg/lost+found/var/clamav/db-20160308T161439Z/main.cvd
/var/pkg/lost+found/var/clamav/db-20160308T161439Z/mirrors.dat
/var/pkg/lost+found/var/clamav/db-20160308T161439Z/bytecode.cvd
/var/pkg/lost+found/var/clamav/run-20160308T161439Z
/var/pkg/lost+found/var/clamav/log-20160309T135657Z
/var/pkg/lost+found/var/clamav/log-20160309T135657Z/freshclam.log
/var/pkg/lost+found/var/clamav/log-20160308T161439Z
/var/pkg/lost+found/var/clamav/log-20160308T161439Z/freshclam.log
/var/pkg/lost+found/var/clamav/run-20160309T135657Z

### Even worse: this leads to a pkg(5) failure unless I explicitly remove the group before reinstalling:
:; pkg install -v clamav
...
DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
Completed                                4/4         58/58      3.1/3.1    0B/s

PHASE                                          ITEMS
Installing new actions                         1/163
Action install failed for 'clamav' (pkg://oi-userland-jim/antivirus/clamav-core):
  KeyError: 'gid'

The Boot Environment hipster_2015.0-20160219T054508Z failed to be updated. A snapshot was taken before the failed attempt and is mounted here /tmp/tmpS30g4A. Use 'beadm unmount hipster_2015.0-20160219T054508Z-2' and then 'beadm activate hipster_2015.0-20160219T054508Z-2' if you wish to boot to this BE.
Traceback (most recent call last):
  File "/usr/bin/pkg", line 5444, in handle_errors
    __ret = func(*args, **kwargs)
  File "/usr/bin/pkg", line 5430, in main_func
    pargs=pargs, **opts)
  File "/usr/bin/pkg", line 2100, in install
    display_plan_cb=display_plan_cb, logger=logger)
  File "/usr/lib/python2.6/vendor-packages/pkg/client/client_api.py", line 1554, in _install
    logger=logger)
  File "/usr/lib/python2.6/vendor-packages/pkg/client/client_api.py", line 1476, in __api_op
    ret = __api_execute_plan(_op, _api_inst)
  File "/usr/lib/python2.6/vendor-packages/pkg/client/client_api.py", line 1028, in __api_execute_plan
    api_inst.execute_plan()
  File "/usr/lib/python2.6/vendor-packages/pkg/client/api.py", line 2825, in execute_plan
    self._img.imageplan.execute()
  File "/usr/lib/python2.6/vendor-packages/pkg/client/imageplan.py", line 4602, in execute
    p.execute_install(src, dest)
  File "/usr/lib/python2.6/vendor-packages/pkg/client/pkgplan.py", line 566, in execute_install
    dest.install(self, src)
  File "/usr/lib/python2.6/vendor-packages/pkg/actions/group.py", line 80, in install
    if (cur_attrs["gid"] != self.attrs["gid"]):
KeyError: 'gid'


pkg: This is an internal error in pkg(5) version ac6c1ce.  Please log a
Service Request about this issue including the information above and this
message.

### So far we must do this:
root@jimoi:~# groupdel clamav
root@jimoi:~# grep clam /etc/{passwd,group,shadow}

### IDEAS WELCOME (Do not provide /var/clamav via pkg? Handle un-installation as a service FMRI?
### What about shut-down alternate roots/zones?)
### How do databases etc. handle it for files they create?
```

Reinstalling and reenabling from clean slate (no guid, no configs):

```
root@jimoi:~# grep clam /etc/{passwd,group,shadow}

root@jimoi:~# pkg install -v clamav
           Packages to install:         4
            Services to change:         1
     Estimated space available:   7.27 GB
Estimated space to be consumed: 641.29 MB  # <<< This seems fun :)
       Create boot environment:        No
Create backup boot environment:        No
          Rebuild boot archive:        No

Changed packages:
oi-userland-jim
  antivirus/clamav
    None -> 0.99.1-2015.0.2.0
  antivirus/clamav-clamdtop
    None -> 0.99.1-2015.0.2.0
  antivirus/clamav-core
    None -> 0.99.1-2015.0.2.0
  antivirus/clamav-milter
    None -> 0.99.1-2015.0.2.0

Services:
  restart_fmri:
    svc:/system/manifest-import:default

DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
Completed                                4/4         58/58      3.1/3.1    0B/s

PHASE                                          ITEMS
Installing new actions                       163/163
Updating package state database                 Done 
Updating package cache                           0/0 
Updating image state                            Done 
Creating fast lookup database                   Done 
Updating package cache                           3/3 

### Stuff created by package installation:
root@jimoi:~# find /var/clamav -ls
   932    2 drwxr-xr-x   5 clamav   clamav          5 Mar  9 14:45 /var/clamav
   952    2 drwxr-xr-x   2 clamav   clamav          2 Mar  9 14:45 /var/clamav/log
   954    2 drwxr-xr-x   2 clamav   clamav          2 Mar  9 14:45 /var/clamav/run
   942    2 drwxr-xr-x   2 clamav   clamav          2 Mar  9 14:45 /var/clamav/db

root@jimoi:~# svcs -a | grep clam
disabled       14:46:21 svc:/antivirus/clamav-milter:default
disabled       14:46:24 svc:/antivirus/clamav:default
disabled       14:46:26 svc:/antivirus/freshclam:default

root@jimoi:~# ls -la /etc/*clam*
ls: cannot access /etc/*clam*: No such file or directory

root@jimoi:~# grep clam /etc/{passwd,group,shadow}
/etc/passwd:clamav:x:6:11:ClamAV Reserved UID:/:
/etc/group:clamav::11:
/etc/shadow:clamav:NL:16869::::::

### Run the provided script to seed the antivirus database and services in this zone:
root@jimoi:~# clamav-enable.sh ; echo $?
INFO: Copying default config '/usr/share/clamav/clamd.conf.sol' into active config '/etc/clamd.conf'
INFO: Copying default config '/usr/share/clamav/freshclam.conf.sol' into active config '/etc/freshclam.conf'
INFO: Copying default config '/usr/share/clamav/clamav-milter.conf.sol' into active config '/etc/clamav-milter.conf'
INFO: Enabling service: FRESHCLAM
ClamAV update process started at Wed Mar  9 14:50:54 2016
Downloading main.cvd [100%]
main.cvd updated (version: 55, sigs: 2424225, f-level: 60, builder: neo)
Downloading daily.cvd [100%]
daily.cvd updated (version: 21455, sigs: 1874555, f-level: 63, builder: neo)
Downloading safebrowsing.cvd [100%]
safebrowsing.cvd updated (version: 44492, sigs: 1993212, f-level: 63, builder: google)
Downloading bytecode.cvd [100%]
bytecode.cvd updated (version: 274, sigs: 49, f-level: 63, builder: anvilleg)
Database updated (6292041 signatures) from db.uk.clamav.net (IP: 81.91.100.173)
WARNING: Clamd was NOT notified: Can't connect to clamd through /var/clamav/run/clamd.sock: No such file or directory
STATE          STIME    FMRI
online         14:54:17 svc:/antivirus/freshclam:default
INFO: Enabling service: CLAMD
svcadm: Instance "svc:/antivirus/clamav:default" is not in a maintenance or degraded state.
STATE          STIME    FMRI
offline*       14:54:22 svc:/antivirus/clamav:default
INFO: Enabling service: CLAMAV-MILTER
svcadm: Instance "svc:/antivirus/clamav-milter:default" is not in a maintenance or degraded state.
STATE          STIME    FMRI
disabled       14:46:21 svc:/antivirus/clamav-milter:default
0

### Ok... reporting was not as nice - might wait a bit before enabling the service to see it starting
root@jimoi:~# svcs -a | grep clam
online         14:54:17 svc:/antivirus/freshclam:default
online         14:54:23 svc:/antivirus/clamav-milter:default
offline*       14:54:22 svc:/antivirus/clamav:default

### And indeed the cronjob is not added:
root@jimoi:~# ls -la /var/spool/cron/crontabs/*cla*
ls: cannot access /var/spool/cron/crontabs/*cla*: No such file or directory

[ Mar  9 14:54:17 Executing start method ("/usr/bin/freshclam.sh ; /usr/bin/freshclam.sh --add-cron "0,15,30,45\ *\ *\ *\ *""). ]
ClamAV update process started at Wed Mar  9 14:54:17 2016
main.cvd is up to date (version: 55, sigs: 2424225, f-level: 60, builder: neo)
daily.cvd is up to date (version: 21455, sigs: 1874555, f-level: 63, builder: neo)
safebrowsing.cvd is up to date (version: 44492, sigs: 1993212, f-level: 63, builder: google)
bytecode.cvd is up to date (version: 274, sigs: 49, f-level: 63, builder: anvilleg)
[ Mar  9 14:54:17 Method "start" exited with status 0. ]

### Redo manually, by copy-pasting the stuff from SMF log...
root@jimoi:~# su - clamav

### maybe escaping is faulty (TODO: explore options)?..
clamav@jimoi:~$ /usr/bin/freshclam.sh ; /usr/bin/freshclam.sh --add-cron "0,15,30,45\ *\ *\ *\ *"
ClamAV update process started at Wed Mar  9 15:02:48 2016
main.cvd is up to date (version: 55, sigs: 2424225, f-level: 60, builder: neo)
daily.cvd is up to date (version: 21455, sigs: 1874555, f-level: 63, builder: neo)
safebrowsing.cvd is up to date (version: 44492, sigs: 1993212, f-level: 63, builder: google)
bytecode.cvd is up to date (version: 274, sigs: 49, f-level: 63, builder: anvilleg)
0,15,30,45\ *\ *\ *\ *  [ -x '/usr/bin/freshclam.sh' ] && '/usr/bin/freshclam.sh'
crontab: error on previous line; unexpected character found in line.
crontab: errors detected in input, no crontab file generated.
ERROR: could not add freshclam to crontab

clamav@jimoi:~$ /usr/bin/freshclam.sh ; /usr/bin/freshclam.sh --add-cron "0,15,30,45 * * * *"    
ClamAV update process started at Wed Mar  9 15:03:01 2016
main.cvd is up to date (version: 55, sigs: 2424225, f-level: 60, builder: neo)
daily.cvd is up to date (version: 21455, sigs: 1874555, f-level: 63, builder: neo)
safebrowsing.cvd is up to date (version: 44492, sigs: 1993212, f-level: 63, builder: google)
bytecode.cvd is up to date (version: 274, sigs: 49, f-level: 63, builder: anvilleg)

clamav@jimoi:~$ ls -la /var/spool/cron/crontabs/*clam*                          
-rw-------   1 root     clamav        78 Mar  9 15:03 /var/spool/cron/crontabs/clamav
-r--------   1 root     clamav        51 Mar  9 15:03 /var/spool/cron/crontabs/clamav.au
```

So indeed... there is some work-up to complete. At least removing the package should be clean...
Suggestions and/or PR's against my repo are welcome though :)

jimklimov commented:
Looking at mariadb component, there is no special magic except making a dummy "common" package to define the UID/GID so they exist before installing/after removing other packages in the set.

On a side note: do we have an authority of some sort to specify the numeric IDs to be reserved for clamav group and user accounts?

pyhalov commented:
There is http://wiki.openindiana.org/oi/UIDs+and+GIDs , you can choose some UID/GID, but remember that a)first user has uid 101, b) we share UID/GID space with illumos-gate.

xen0l commented:
Is there anything else that needs to be done here?

jimklimov commented:
Bumping recipe since a new clamav version came out recently (old ones will complain into logs when downloading virus DB definitions).

jimklimov commented:
Also grabbed definite GID/UID for clamav (via Wiki)

jimklimov commented:
Also found that C++ bindings were not building due to recipe issues

jimklimov commented:
xml2-config 32-bit version is always executed causing warnings in 64-bit .so lib
tinkering on a workaround

jimklimov commented:
Fixed usage of external *-config scripts. Oh boy is their implementation lousy and dirty! ;-\
Also moved into under sysutils/clamav directory.
Should be ok to merge now, I don't plan further improvements in this PR.

xen0l commented:
lgtm

pyhalov commented on components/sysutils/clamav/clamav-core.p5m:
```diff
@@ -0,0 +1,138 @@
+    restart_fmri=svc:/system/manifest-import:default
+file \
+    files/freshclam.xml \
+    path=var/svc/manifest/site/freshclam.xml \
```
please, deliver services in lib/svc/manifest/application , to be consistent with other system services

pyhalov commented:
[ May  7 23:20:22 Executing start method ("/usr/bin/freshclam.sh ; /usr/bin/freshclam.sh --add-cron 0,15,30,45\ *\ *\ *\ *"). ]
[ May  7 23:20:22 Method "start" exited with status 1. ]

pyhalov commented:
Perhaps, it's worth just make separate method script and use it with simple parameters start/stop, and script will call freshclam.sh with necessary arguments? This could make space escaping issues more simple.

jimklimov commented:
The "simple script" would sort of defeat the trumpeted big cause of moving config bits into SMF, away from config files...

pyhalov commented:
No, it still can reference SMF properties. But you shouldn't worry so much about symbol escaping,
Perhaps, even clamav-enable.sh logic can be moved there

jimklimov commented:
Ok, pardon my late-night comment. Of course we can read the svcprop from a method script too. :-\
For now, added REQUIRED_PACKAGES to the PR.

jimklimov commented:
Script+XML for freshclam are fixed^H^H^H^H^Hchanged now.

pyhalov commented on components/sysutils/clamav/Makefile:
```diff
@@ -0,0 +1,175 @@
+GNULIBDIR.32=$(GNULIB)
+GNULIBDIR.64=$(GNULIB64)
+
+LDFLAGS.$(BITS) +=-L$(GNULIBDIR.$(BITS)) -R$(GNULIBDIR.$(BITS)) -L$(USRLIBDIR.$(BITS)) -R$(USRLIBDIR.$(BITS))
```
Again, last -L -R shouldn't be necessary

pyhalov commented:
svcadm enable freshclam...
tail -f /var/svc/log/antivirus-freshclam:default.log:

0,15,30,45\ *\ *\ *\ \*  [ -x '/usr/bin/freshclam.sh' ] && '/usr/bin/freshclam.sh'
crontab: error on previous line; unexpected character found in line.

 svcprop freshclam |grep cron-s
freshclam/cron-schedule astring 0,15,30,45\ *\ *\ *\ *

jimklimov commented:
Hope this should be fixed now.
